src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -215,30 +215,33 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
                 // absolute line index to use later when combining results.
                 var absoluteLine = 0;
                 var isFirstTokenInRange = true;
-                for (var tokenIndex = 0; tokenIndex < startingTokens.Data.Length; tokenIndex++)
+                for (var tokenIndex = 0; tokenIndex < startingTokens.Data.Length; tokenIndex += TokenSize)
                 {
                     // The first int of each token represents the line offset. We use this info to keep track of the
                     // absolute line for use later on when computing relative token positions.
-                    if (tokenIndex % TokenSize == 0)
+                    absoluteLine += startingTokens.Data[tokenIndex];
+
+                    // Skip tokens that are out of the range
+                    if (absoluteLine < requestedRange.Start.Line || absoluteLine >= cachedRange.Start.Line)
+                    {
+                        continue;
+                    }
+
+                    // First token of the starting set should be relative to the start of the document
+                    if (isFirstTokenInRange)
+                    {
+                        finalTokens.Add(absoluteLine);
+                        isFirstTokenInRange = false;
+                    }
+                    else
                     {
-                        absoluteLine += startingTokens.Data[tokenIndex];
-
-                        // Skip tokens that are out of the range
-                        if (absoluteLine < requestedRange.Start.Line || absoluteLine >= cachedRange.Start.Line)
-                        {
-                            continue;
-                        }
-
-                        // First token of the starting set should be relative to the start of the document
-                        if (isFirstTokenInRange)
-                        {
-                            finalTokens.Add(absoluteLine);
-                            isFirstTokenInRange = false;
-                            continue;
-                        }
+                        finalTokens.Add(startingTokens.Data[tokenIndex]);
                     }
 
-                    finalTokens.Add(startingTokens.Data[tokenIndex]);
+                    finalTokens.Add(startingTokens.Data[tokenIndex + 1]);
```

Comment:
probably better to do a nested for loop here 

```suggestion
					foreach (var i = 0; i < TokenSize; i++)
					{
						finalTokens.Add(stringTokens.Data[tokenIndex + i];
					}
```

Reply:
Ohh, good idea @ryzngard!

Reply:
Ohh that does clean things up, thanks!